File: src/main/java/me/ksviety/teleporter/teleporters/OneTimeEntityTeleporter.java
package me.ksviety.teleporter.teleporters;

import me.ksviety.teleporter.EntityTeleporter;
import me.ksviety.teleporter.cache.TeleportationCache;
import net.minecraft.entity.Entity;

import java.util.Collection;

public class OneTimeEntityTeleporter implements EntityTeleporter {
    private final EntityTeleporter entityTeleporter;
    private final TeleportationCache cache;

    public OneTimeEntityTeleporter(EntityTeleporter entityTeleporter, TeleportationCache cache) {
        this.entityTeleporter = entityTeleporter;
        this.cache = cache;
    }

    @Override
    public void teleport(Entity entity) {
        final String name = entity.getName();
        final Collection<String> teleported = cache.getAlreadyTeleportedPlayers();

        if (teleported.contains(name))
            return;

        entityTeleporter.teleport(entity);
        cache.addPlayerAsTeleported(name);
    }
}

File: src/main/kotlin/me/ksviety/teleporter/teleporters/OneTimeEntityTeleporter.kt
package me.ksviety.teleporter.teleporters

import me.ksviety.teleporter.EntityTeleporter
import me.ksviety.teleporter.Cache
import net.minecraft.entity.Entity

class OneTimeEntityTeleporter(
    private val entityTeleporter: EntityTeleporter,
    private val cache: Cache
    ) : EntityTeleporter {

    override fun teleport(entity: Entity) {
        val name = entity.name
        val teleported = cache.alreadyTeleportedPlayers

        if (teleported.contains(name))
            return

        entityTeleporter.teleport(entity)
        cache.addPlayerAsTeleported(name)
    }
}
